### DISCUSSION TITLE
Most important features for the AvaloniaVS project

### DISCUSSION THREAD

**Discussion by Takoooooo** [AvaloniaUI/AvaloniaVS#417] [General]
We are wanting to continue enhancing the integration of the Avalonia into VS and currently we want to understand what features the community needs most. Please write in the comments which features you would like to see the most and like the comments of other community members which you find important ;)

**Comment by vdtoorn** [comment]
It would be useful to see the errors near the "stop sign" on top of the designer instead of having to switch to the output window all the time.

**Comment by nsrahmad** [comment]
![Screenshot_20231115_104006](https://github.com/AvaloniaUI/AvaloniaVS/assets/10996250/49c7dcca-4643-4091-b9ed-db9cc1c2b23d)

The way Rider hints at Binding Path is really helpful as a beginner. I wish VS plugin could do something similar.

**Comment by laolarou726** [comment]
Currently VS cannot recognize most of Keys from my StyleIncludes. For example when I using DynamicResource/StaticResource, I have to copies those keys from their source Adam's instead of using Intellisense.

**Comment by strangerinla** [comment]
Need more integration with view model:
- Go to Definition of bound view model property by F12
- View model properties auto-complete

**Comment by r9guy** [comment]
Can we have  the Xaml line numbering feature?

**Comment by r9guy** [comment]
it would be really nice to have a toolbox with drag and drop controls.
I

**Reply by Daniil-Kolbasov** [reply]
When I first started learning UWP toolbox helped me a lot. I would like to see something similar in Avalonia

**Reply by Takoooooo** [reply]
Not planned, hiding

**Comment by blue00** [comment]
The backward and forward navigation buttons do not work inside .xaml or .axaml files.

I like to navigate back and forth between my last focused code lines a lot. That's why this would be an important feature for me.

![navbuttons](https://github.com/AvaloniaUI/AvaloniaVS/assets/8476361/091d1569-6c11-4734-881b-b1388ddc58a3)

**Reply by grantbowering** [reply]
yeah, I `Ctrl`+`-` a lot and it just skips over .axaml files and goes to the previous .cs file.  That's in my top pet peeves for sure

**Comment by blue00** [comment]
Easier previewer debuggig would be nice. The exception in the Avalonia Diagnostics output was often not enough for me to find out why the preview is broken.

At the moment you have to find out the process ID and then attach the debugger manually.

In some cases you also have to add these lines to Program.cs to make it work:
```csharp
Debugger.Launch();
while (!Debugger.IsAttached)
    Thread.Sleep(100);`
 ```
(Source: https://docs.avaloniaui.net/docs/0.10.x/guides/developer-guides/debugging-previewer)

**Comment by jpgarza93** [comment]
1. HotReload would be a game changer.
2. XAML intellisense/auto-complete. WPF does a great job showing you all DynamicResources. With Avalonia I spend a lot of time finding DynamicSources because I need to navigate to code.
3. Improve xaml errors. Sometimes it’s very hard to tell why the designer is not rendering. Most of the times it’s a small xaml error. But you have to find it yourself.

**Comment by grantbowering** [comment]
I think my single biggest pet peeve is TextBlock literal child text being red-underlined in the XAML pane of the Avalonia XAML Editor and red-ticked in its scrollbar.

![image](https://github.com/AvaloniaUI/AvaloniaVS/assets/2152220/095b8d93-515f-4065-94e5-d0d08c63fd25)

I mean, it design previews fine, it builds and runs fine, but it's just incorrect noise that makes me twitch and can make any real problems harder to find.

**Reply by strangerinla** [reply]
There are also formatter problems:
It formats the string `<Run Text="123"/><Run Text="456"/>` into separate lines which results in extra space between runs in output. 
It would also be cool to have hyperlink.

**Reply by Takoooooo** [reply]
Can't repro. Open an issue if you know how to repro it. I assume in your cse the problem might be in outdted Avloni extension version.
![image](https://github.com/AvaloniaUI/AvaloniaVS/assets/53405089/12ab3cd1-127d-438d-9050-1e5d864b2b1c)

**Comment by AtomCrafty** [comment]
One trap I find myself falling for again and again is that the previewer uses the `*.Desktop` project, but pressing Ctrl+B while inside an axaml file will only rebuild the shared project. Way too many times have I tried to figure out why the previewer complains about non-existent classes and properties, only to realize that the desktop project was not up to date. Ctrl+Shift+B works around the issue, but that can be slow in solutions with many projects.

**Comment by grantbowering** [comment]
Something I really miss from the WPF XAML editor is that when you had it in split mode and you clicked on an element in the preview pane it would scroll to the markup for that element in the xaml pane.

**Reply by jpgarza93** [reply]
Forgot about this one, great feature 👍

**Comment by chenzanyu** [comment]
Although not the most important, I think the default template item should be upgraded to .NET8

**Reply by Takoooooo** [reply]
Updated https://github.com/AvaloniaUI/avalonia-dotnet-templates/pull/241

**Comment by WilbertOnGithub** [comment]
I don't know if this is the correct place to mention this (if not, please move or delete and accept my apologies) but the main thing I was stumbling with was having a decent default template. With this I mean the following:

- Something that can be installed with 'dotnet new'.
- Has a sufficiently fleshed out example (instead of only hello world).
- Shows MVVM usage combined with all the current best practices
- Usage of the MVVM Community Toolkit
- Shows how to do Dependency Injection for your views/viewmodels using the DI framework provided by Microsoft.

I'm working on a hobby project (https://github.com/WilbertOnGithub/ParkingBarrierManager) that I intend to use internally at my work as a showcase. 

If necessary, this could be the basis for a new template. Let me know if anyone is interested. Never wrote a template before so that will be something to figure out.

**Comment by Daniil-Kolbasov** [comment]
I would like to see Navigator as in Affinity for the convenience of navigating the page with zoom

![Group](https://github.com/AvaloniaUI/AvaloniaVS/assets/34882813/44a451b0-abfa-4316-9387-fa04c0f71761)
